### ISSUE TITLE
IREP - fix number of rules

### ISSUE THREAD

**Issue by abhishek-ghose** [imoscovitz/wittgenstein#3]
Hi,

I recently came across your library, and to begin with, thank you for writing this!
I have a question regarding the usage of the `IREP` technique: can I limit the number of rules that the ruleset grows to? Similar to the `max_depth` parameter decision tree libraries usually provide. I took a quick look at the code, and I think this is not currently available. Wanted to confirm.

Thank you!

**Comment by imoscovitz** [comment]
Hi abhishek,

Thanks for writing!

I had been thinking of implementing that option -- now that you've requested it, I'll get on it :) I'll be publishing a small version update coming soon to allow predict_proba. I can include a max_length feature into the update.

Two questions for you:
- How large is the model you're getting, in conds and in rules? To check conds, call clf.ruleset_.count_conds(). To check rules, call len(clf.ruleset_)
- Is the reason you want to limit the number of model size to have a more interpretable model, or to cut down on training time?

Thanks,
Ilan

**Comment by abhishek-ghose** [comment]
Thank you for responding!

Here are my answers:
- 46 conditions and 24 rules
- primarily for interpretability. My use-case is something like "I want to see what's the best the system can do with `n` rules, since `n` is all I can parse"

**Comment by imoscovitz** [comment]
Thanks Abhiskek!

- I've implemented three parameters: max_rules, max_rule_conds, max_total_conds for both IREP and RIPPER. That way it's possible to control how the interpretation is structured. Their implementation is designed to improve interpretability and training speed, rather than specifically model performance, since model training already includes pruning. But tuning them like hyperparameters might work anyways -- the dataset I happened to be testing RIPPER's code on saw a large (~10 points) accuracy improvement from a slow, difficult-to-interpret model to a fast, interpretable model. Full descriptions are in the docstrings of the clf.fit method.

- Let me know if you find it helpful! I committed the changes to this repo just in case you'd like to try it out before I publish the new version this week -- also, that way in case you have any feedback on how it's working for you, I'd be able to include it into the update.
- Also, I'd be really curious to hear what domain you're using it for. I saw your github has a repo on text mining, and I've for awhile been contemplating how the library might be be applied to NLP or deep learning.

Thanks again!
Ilan

**Comment by abhishek-ghose** [comment]
Hi,

Thanks for the very quick turnaround! These are indeed very helpful options. I tried varying the `max_rules` in `IREP` for the [higgs](https://www.csie.ntu.edu.tw/~cjlin/libsvmtools/datasets/binary.html) dataset. Here is what I see:

![image](https://user-images.githubusercontent.com/1311145/55285128-30983200-533a-11e9-9461-30155ea401ea.png)

My github account is more or less defunct :-) - most of my stuff is in private repos on bitbucket. I am trying out a bunch of interpretable classifiers to test out some stuff for my PhD research. I will shortly (within 2 weeks) put up something on arxiv. 

Coming to use of rules in NLP, one use case seems quite obvious. If you represent documents as a BoW, a ruleset (used as a classifier) gives you an interpretable way to see the presence or absence of which words indicate a particular class.

**Comment by imoscovitz** [comment]
Thanks, Abhishek!

- the `max_rules`/actual disparity is strange. Did you pass a random_state parameter to the .fit() method? Internal to the .fit() training algorithm is a pruning stage that makes randomized train/prune splits which can affect the model and model length it produces. Let me know if that fixes the problem for you or if you've already tried that.

- On a related note that your graph also illustrates, because a ruleset is a set of OR's, (assuming all else is equal, random_states, etc.), a higher max_rules will _always_ produce at least the same or better recall, and lower max_rules is _likely_ (rather than always, I believe) to produce the same or better precision.

- I had actually played around a little bit with BoW as a starting point. Any thoughts on strategies for reducing the dimensionality/vocab size without losing interpretability? There's obviously topic modeling, but that's starting to lose the advantages of rules. Would you suggest (after basic preprocessing/stopwords etc.) setting max and min frequencies as the main strategy, or something else/in addition?

- Excited to see your arxiv! Will you send it to me once it's up?

Thanks again! In the meantime, let me know if there's anything else I can do to help with the package.

Ilan

**Comment by abhishek-ghose** [comment]
Hi Ilan,

Thank you for your response!
I believe `random_state` indeed was the culprit. Same experiment with `random_state` fixed leads me to this graph:

![image](https://user-images.githubusercontent.com/1311145/55457078-b962d800-559d-11e9-9925-dc9c0b0d4315.png)

This time around all metrics - *precision*, *recall*, *F1* - exactly converge. Which is how it should be - beyond a certain value for `max_rules`, the number of rules actually learned must remain the same. NB: I had fixed `max_rule_conds=3` too - in this experiment and the one before, to make sure that number of rules is not compensated for by learning larger rules.

Here's what the `max_rules` and actual rules look like now:

- `max_rules=1`, actual=1
- `max_rules=3`, actual=3
- `max_rules=5`, actual=5
- `max_rules=7`, actual=7
- `max_rules=9`, actual=9
- `max_rules=11`, actual=9
- `max_rules=13`, actual=9
- `max_rules=15`, actual=9

Regarding the BoW question, nothing else strikes me immediately.

Sure, will let you know once the article is up!

**Comment by imoscovitz** [comment]
Hi abhishek,

Great! So glad that helped. That really is a beautiful graph, and it's causing me me think about how to visualize the stuff below, perhaps in the Readme.

> NB: I had fixed max_rule_conds=3 too - in this experiment and the one before, to make sure that number of rules is not compensated for by learning larger rules.

That's a good idea to keep other params constant during that experiment. Just a thought about what could be going on, because their interaction could be a little surprising :)

1. Longer rules contribute to precision by _tightening_ what a rule will predict positive; a greater number of rules contributes to recall by saying, "here is another description you can use for what positive examples look like", thus _loosening_ the final model. So rule length and and number of rules contribute to the model in opposite ways. 

2. The way the implementation works, pruning happens after each rule is built, but once you move on to the next rule there isn't look-back (at least for IREP -- RIPPER is a bit different). So `max_rule_conds` can affect the number of rules produced, but `max_rules` probably can't affect the size of each rule. (Again, for IREP.) 

Combine 1) and 2) and what I suspect might be happening, is that limiting `max_rules` on IREP shouldn't cause it to compensate by making more complex rules; rather limiting `max_rule_conds` could very well cause it to compensate by actually _reducing_ complexity even further by limiting the optimal number of rules. It's possible limiting `rule_conds` to 3 could be causing the optimal number of rules to peter out around 9, because it could be saying, "the rules I made so far have been permissive (because they're short); their permissiveness means recall is going great -- let's compensate by also shortening the number of rules in order to protect precision. (Which is kind of cool and hilarious.)

Shorter rules causing fewer rules is really different from how we're used to thinking about hyperparameters in other types of models, where their reaction to simplicity in one area is to add complexity in another. But it might not be a bad thing because if interpretability is the goal, getting one kind of simplicity as a side-effect of another simplicity could be desirable for many domains.

I'll keep experimenting with and asking around about BoW feature selection.

Really looking forward to your article! 

Thanks again,
-Ilan

**Comment by abhishek-ghose** [comment]
Sorry for dropping off the radar!
Yes, I agree with you that individual rules (and their conditions) increase precision, since the ruleset is a DNF, and more rules loosen the definition of a class. We do see this in the plot above - the recall increases as the number of rules increase.

But to your other point, I hadn't really thought about it before, but it makes sense. I ran a quick test (**NOTE - overfitting alert!** - only used a training set, no validation set!) - I fixed the `max_rules=20` and tried various values of `max_rule_conds`. This is what I see:

![image](https://user-images.githubusercontent.com/1311145/56555123-bb8ad780-6548-11e9-8694-fdb45edfeb12.png)

So, yeah :)

Another question - does this handle binary class problems only now? I skimmed through the code, and I understand that the first label in the training set is assumed to be the positive class, and the rest is negative. Is that right? (Its easy enough to extend using *scikit's* `OneVsRestClassifier`, but wanted to confirm before I did so).

Thanks!

**Comment by imoscovitz** [comment]
Hi @abhishek-ghose, great to hear from you! No worries, you must have been busy -- how's your paper going?

- Wow. That is a really cool graph! What's your metric -- auc?
- Just so you know in case it's helpful for you, there are also now major speed optimizations for both IREP and RIPPER committed to the repo. Here's a fairly representative anecdote -- the commit is about one order of magnitude faster for IREP. https://github.com/imoscovitz/wittgenstein/issues/4 There is also now a predict_proba
- Yep -- you got it right. Currently it's only for binary class because I was working on other things and thinking people would use OneVsRestClassifier. I'm not sure, but I believe IREP will only predict one class and RIPPER has the possibility of predicting more than one, but I'd have to think about it. If you want just one class, you can select the one whose class frequency is the highest, or try using the new experimental predict_proba.
- I was thinking of doing a multi-class implementation. It would essentially be OneVsRest. Pros: it wouldn't be that difficult to create, and users would be less reliant on sklearn; Cons: there might be more useful priorities to focus on, and it could involve a little refactoring and/or a little tweaking to .fit parameter usage. Do you have an opinion on whether it is worth doing?

Thanks!
Ilan

**Comment by abhishek-ghose** [comment]
Hi @imoscovitz ,

Thanks! I was using the F1 score.

I noticed the check-ins. And just now took a look at the speedup table - the improvement is great. Here's my embarrassing little secret though - my codebase (which has been around for a while now) is in Python 2.7, and the way I have been using your IREP code is by - *gasp* - converting them. Since that works now, it'll be a while before I get to the sped up version though.

I think scikit's `OneVsRestClassifier` simple enough to use; one pro in rolling out your own would be the control on parameters like `max_rules`: is this enforce across classes, or per one-vs-rest classifier.

The paper is still a while off, unfortunately, slowly getting there, most likely in the 1st week of May.

**Comment by abhishek-ghose** [comment]
Hi @imoscovitz 

[This](https://arxiv.org/abs/1905.01520 ) is what I've been working on.
Unfortunately I couldn't get results from IREP in  - ran out of time to make a submission, I hope you still find it an interesting read!

Regards

**Comment by shawn-jung** [comment]
Thank you for this great conversation you had in 2019! Speaking of this cool `max_rules` parameter, is it fair to say the rules are prioritized that explains or covers lots of data? (or say, by the order of mutual information something like that?)
